packages/flutter_tools/test/general.shard/devfs_test.dart

```diff
@@ -332,6 +332,74 @@ void main() {
       Uri.parse('goodbye'): DevFSFileContent(file),
     }, Uri.parse('/foo/bar/devfs/')), throwsA(isA<DevFSException>()));
   });
+
+  testWithoutContext('test handles request closure hangs', () async {
+    final FileSystem fileSystem = MemoryFileSystem.test();
+    final FakeVmServiceHost fakeVmServiceHost = FakeVmServiceHost(
+      requests: <VmServiceExpectation>[createDevFSRequest],
+    );
+    final HttpClient httpClient = MockHttpClient();
+    final MockHttpClientRequest httpRequest = MockHttpClientRequest();
+    when(httpRequest.headers).thenReturn(MockHttpHeaders());
+    when(httpClient.putUrl(any)).thenAnswer((Invocation invocation) {
+      return Future<HttpClientRequest>.value(httpRequest);
+    });
+    final MockHttpClientResponse httpClientResponse = MockHttpClientResponse();
+    bool didAbort = false;
+    bool didRetry = false;
+    when(httpRequest.close()).thenAnswer((Invocation invocation) async {
+      if (!didAbort) {
+        while (!didAbort) {
+          await new Future.delayed(const Duration(milliseconds : 10));
+        }
+        throw HttpException('aborted');
+      }
+      didRetry = true;
+      return httpClientResponse;
+    });
+    when(httpRequest.abort()).thenAnswer((_) {
+      didAbort = true;
+    });
```

Comment:
We can use completers to avoid the polling loop. This also uses a completer for the result of close(), so that the same Future can be returned for `httpRequest.done`.

```dart
    final Completer<void> abortCompleter = Completer<void>();
    final Completer<HttpClientResponse> responseCompleter = Completer<HttpClientResponse>();
    when(httpRequest.close()).thenAnswer((Invocation invocation) async {
      if (!didAbort) {
        await abortCompleter.future;
        unawaited(Future<void>(() async {
          responseCompleter.completeError(HttpException('aborted'));
        });
        return responseCompleter.future;
      }
      didRetry = true;
      return httpClientResponse;
    });
    when(httpRequest.abort()).thenAnswer((_) {
      didAbort = true;
      abortCompleter.complete();
    });
    when(httpRequest.done).thenReturn(responseCompleter.future);
```

Reply:
Done